Testnet/NFTAuctionStore-Royalties/NFTAuctionStore/NFTAuctionStore.cs

```diff
@@ -0,0 +1,293 @@
+    {
+        var auction = GetAuctionInfo(contract, tokenId);
+
+        Assert(auction.OnAuction, "Auction is not found.");
+
+        Assert(!EndBlockReached(auction), "Auction ended.");
+
+        Assert(Message.Value > auction.HighestBid && Message.Value >= auction.StartingPrice, "The amount is not higher than highest bidder or starting price.");
+
+        if (auction.HighestBid > 0)
+        {
+            //refund for previous bidder
+            var balance = GetRefund(auction.HighestBidder) + auction.HighestBid;
+
+            if (!State.IsContract(auction.HighestBidder))
+            {
+                var result = Transfer(auction.HighestBidder, balance);
+
+                if (result.Success)
+                    balance = 0;
+            }
+
+            SetRefund(auction.HighestBidder, balance);
+        }
+
+        auction.HighestBidder = Message.Sender;
+        auction.HighestBid = Message.Value;
+
+        SetAuctionInfo(contract, tokenId, auction);
+
+        Log(new HighestBidUpdatedLog { Contract = contract, TokenId = tokenId, Bidder = auction.HighestBidder, Bid = auction.HighestBid });
+    }
+
+    private bool EndBlockReached(AuctionInfo auction)
+    {
+        return Block.Number >= auction.EndBlock;
+    }
+
+    public bool Refund()
+    {
+        EnsureNotPayable();
+
+        var amount = GetRefund(Message.Sender);
+
+        Assert(amount > 0);
+
+        SetRefund(Message.Sender, 0);
+
+        var transfer = Transfer(this.Message.Sender, amount);
+
+        Assert(transfer.Success, "Transfer failed.");
+
+        Log(new BalanceRefundedLog { To = Message.Sender, Amount = amount });
+
+        return transfer.Success;
+    }
+
+    public void AuctionEnd(Address contract, UInt256 tokenId)
+    {
+        EnsureNotPayable();
+
+        var auction = GetAuctionInfo(contract, tokenId);
+
+        Assert(auction.OnAuction, "Auction is not found.");
+
+        Assert(EndBlockReached(auction), "Auction is not ended yet.");
+
+        auction.OnAuction = false;
+        SetAuctionInfo(contract, tokenId, auction);
+
+        if (auction.HighestBid == 0)
+        {
+            TransferToken(contract, tokenId, Address, auction.Seller);
+
+            Log(new AuctionEndFailedLog { Contract = contract, TokenId = tokenId });
+            return;
+        }
+
+        var royalty = GetRoyaltyInfo(contract, tokenId, auction.HighestBid);
+
+        var highestBidMinusRoyalty = auction.HighestBid - royalty.Amount;
+
+        if (State.IsContract(auction.Seller))
+        {
+            var balance = GetRefund(auction.Seller);
+            SetRefund(auction.Seller, balance + highestBidMinusRoyalty);
+        }
+        else
+        {
+            var result = Transfer(auction.Seller, highestBidMinusRoyalty);
+
+            Assert(result.Success, "Transfer failed.");
+        }
+
+        if (royalty.Amount > 0)
+        {
+            var royaltyTransfer = Transfer(royalty.Recipient, royalty.Amount);
```

Comment:
This probably should have the same treatment as paying `auction.Seller` where if it's a contract, it sets a refund amount to be collected via pull. 

Resolves edge case where `royalty.Recipient` is a contract that is not payable, locking the seller's CRS in the contract with no way to end the auction.